File: 3121-count-the-number-of-special-characters-ii/3121-count-the-number-of-special-characters-ii.py
class Solution:
    def numberOfSpecialChars(self, word: str) -> int:
        last_idx_lower = [-1] * 26
        first_idx_upper = [-1] * 26

        for i, c in enumerate(word):
            if c.islower():
                last_idx_lower[ord(c) - ord('a')] = i
            elif first_idx_upper[ord(c) - ord('A')] == -1:
                first_idx_upper[ord(c) - ord('A')] = i

        count = 0
        print(last_idx_lower, first_idx_upper)
        for j in range(26):
            if last_idx_lower[j] != -1 and first_idx_upper[j] != -1:
                count += last_idx_lower[j] < first_idx_upper[j]

        return count

        

File: 3121-count-the-number-of-special-characters-ii/3121-count-the-number-of-special-characters-ii.java
class Solution {
    public int numberOfSpecialChars(String word) {
        int[] lastIdxLower = new int[26];
        int[] firstIdxUpper = new int[26];

        Arrays.fill(lastIdxLower, -1);
        Arrays.fill(firstIdxUpper, -1);

        for (int i=0; i < word.length(); i++){
            char c = word.charAt(i);

            if (Character.isLowerCase(c)){
                lastIdxLower[c - 'a'] = i;
            } else if (firstIdxUpper[c - 'A'] == -1){
                firstIdxUpper[c - 'A'] = i;
            }
        }

        int count = 0;

        for (int j = 0; j < 26; j++){
            if (lastIdxLower[j] != -1 && lastIdxLower[j] < firstIdxUpper[j]){
                count ++;
            }
        }

        return count;
    }
}